File: packages/server/src/gate.ts
import { Hono } from "hono";
import { z } from "zod";
import {
  ErrorCodes,
  ToolgateError,
  decide,
  hashArgs,
  newId,
  verifyCapabilityToken,
  verifyPopProof,
  type ApprovalRequest,
  type AuditDecision,
  type CapabilityClaims,
  type DelegationGrant,
  type ToolDef,
  type Upstream,
} from "@toolgate/core";
import type { AppContext } from "./context.js";

const CallBodySchema = z.object({
  tool: z.string().min(1),
  args: z.record(z.string(), z.unknown()).default({}),
});

interface AuthedCall {
  claims: CapabilityClaims;
  grant: DelegationGrant;
}

export function gateRoutes(ctx: AppContext): Hono {
  const app = new Hono();

  app.post("/call/:upstream", async (c) => {
    const upstreamName = c.req.param("upstream");
    const rawBody = await c.req.json().catch(() => {
      throw new ToolgateError(ErrorCodes.VALIDATION, "request body must be JSON");
    });
    const parsed = CallBodySchema.safeParse(rawBody);
    if (!parsed.success) {
      throw new ToolgateError(ErrorCodes.VALIDATION, "invalid call body", {
        issues: parsed.error.issues.map((i) => `${i.path.join(".")}: ${i.message}`),
      });
    }
    const { tool, args } = parsed.data;

    const { claims, grant } = await authenticate(ctx, c.req.raw, `/v1/gate/call/${upstreamName}`);
    const { upstream, toolDef } = resolveTool(ctx, claims.tenant, upstreamName, tool);

    const callId = newId("call");
    const call = { upstream: upstreamName, tool, args, costUnits: toolDef.costUnits };
    const policy = ctx.store.getPolicy(grant.policyId);
    if (!policy) throw new ToolgateError(ErrorCodes.INTERNAL, "grant policy missing");

    const decision = decide(policy, call, claims.authorization_details);
    const auditBase = {
      id: newId("evt"),
      tenantId: claims.tenant,
      ts: new Date().toISOString(),
      actor: {
        agentId: claims.act.sub,
        userId: claims.sub,
        grantId: grant.id,
        tokenJti: claims.jti,
      },
      action: { callId, upstream: upstreamName, tool, argsHash: hashArgs(args) },
    };

    if (decision.effect === "deny") {
      ctx.audit.record({
        ...auditBase,
        decision: toAuditDecision(decision),
        result: { status: "denied" },
      });
      throw new ToolgateError(ErrorCodes.DENIED, decision.reason, {
        source: decision.source,
        ...(decision.ruleId ? { ruleId: decision.ruleId } : {}),
      });
    }

    if (decision.effect === "require_approval") {
      const approval: ApprovalRequest = {
        id: newId("apr"),
        tenantId: claims.tenant,
        callId,
        grantId: grant.id,
        agentId: claims.act.sub,
        userId: claims.sub,
        upstream: upstreamName,
        tool,
        args,
        status: "pending",
        requestedAt: new Date().toISOString(),
        expiresAt: new Date(Date.now() + ctx.config.approvalTtlSeconds * 1000).toISOString(),
      };
      ctx.store.putApproval(approval);
      ctx.audit.record({
        ...auditBase,
        decision: toAuditDecision(decision),
        result: { status: "pending_approval" },
      });
      return c.json(
        {
          status: "pending_approval",
          approval_id: approval.id,
          expires_at: approval.expiresAt,
          reason: decision.reason,
        },
        202,
      );
    }

    const result = await executeCall(ctx, {
      auditBase,
      decision: toAuditDecision(decision),
      grant,
      upstream,
      toolDef,
      tool,
      args,
    });
    return c.json({ status: "executed", call_id: callId, result });
  });

  app.get("/approvals/:id", async (c) => {
    const { claims } = await authenticateTokenOnly(ctx, c.req.raw);
    const approval = requireApproval(ctx, c.req.param("id"), claims);
    return c.json({
      approval_id: approval.id,
      status: approval.status,
      expires_at: approval.expiresAt,
    });
  });

  app.post("/approvals/:id/execute", async (c) => {
    const approvalId = c.req.param("id");
    const { claims, grant } = await authenticate(
      ctx,
      c.req.raw,
      `/v1/gate/approvals/${approvalId}/execute`,
    );
    const approval = requireApproval(ctx, approvalId, claims);

    if (approval.status === "pending") {
      throw new ToolgateError(ErrorCodes.APPROVAL_PENDING, "approval still pending");
    }
    if (approval.status !== "approved") {
      throw new ToolgateError(ErrorCodes.APPROVAL_DENIED, `approval is ${approval.status}`);
    }
    if (new Date(approval.expiresAt).getTime() < Date.now()) {
      approval.status = "expired";
      ctx.store.putApproval(approval);
      throw new ToolgateError(ErrorCodes.APPROVAL_DENIED, "approval expired before execution");
    }

    const { upstream, toolDef } = resolveTool(ctx, claims.tenant, approval.upstream, approval.tool);
    // The stored args are the approved args — the agent cannot substitute them.
    const auditBase = {
      id: newId("evt"),
      tenantId: claims.tenant,
      ts: new Date().toISOString(),
      actor: {
        agentId: claims.act.sub,
        userId: claims.sub,
        grantId: grant.id,
        tokenJti: claims.jti,
      },
      action: {
        callId: approval.callId,
        upstream: approval.upstream,
        tool: approval.tool,
        argsHash: hashArgs(approval.args),
      },
    };
    const decision: AuditDecision = {
      effect: "allow",
      source: "approval",
      reason: `approved by ${approval.decidedBy ?? "unknown"} at ${approval.decidedAt ?? "?"}`,
    };

    const result = await executeCall(ctx, {
      auditBase,
      decision,
      grant,
      upstream,
      toolDef,
      tool: approval.tool,
      args: approval.args,
    });

    approval.status = "executed";
    approval.executedAt = new Date().toISOString();
    ctx.store.putApproval(approval);
    return c.json({ status: "executed", call_id: approval.callId, result });
  });

  return app;
}

// ---------------------------------------------------------------------------
// Pipeline stages
// ---------------------------------------------------------------------------

async function authenticateTokenOnly(ctx: AppContext, req: Request): Promise<AuthedCall> {
  const auth = req.headers.get("authorization") ?? "";
  if (!auth.toLowerCase().startsWith("bearer ")) {
    throw new ToolgateError(ErrorCodes.TOKEN_INVALID, "missing bearer capability token");
  }
  const token = auth.slice(7).trim();
  const claims = await verifyCapabilityToken(ctx.keys.control.publicJwk, token, {
    issuer: ctx.config.issuer,
    audience: ctx.config.gateAudience,
  });

  const grant = ctx.store.getGrant(claims.grant_id);
  if (!grant) throw new ToolgateError(ErrorCodes.NOT_FOUND, "grant no longer exists");
  if (grant.status !== "active") {
    // Live tokens die with their grant: revocation is immediate.
    throw new ToolgateError(ErrorCodes.REVOKED, "grant revoked");
  }
  if (new Date(grant.expiresAt).getTime() < Date.now()) {
    throw new ToolgateError(ErrorCodes.TOKEN_EXPIRED, "grant expired");
  }
  const agent = ctx.store.getAgent(claims.act.sub);
  if (!agent || agent.status !== "active") {
    throw new ToolgateError(ErrorCodes.REVOKED, "agent unknown or disabled");
  }
  return { claims, grant };
}

async function authenticate(ctx: AppContext, req: Request, path: string): Promise<AuthedCall> {
  const authed = await authenticateTokenOnly(ctx, req);
  const proof = req.headers.get("x-toolgate-proof");
  if (!proof) throw new ToolgateError(ErrorCodes.PROOF_INVALID, "missing x-toolgate-proof header");

  const token = (req.headers.get("authorization") ?? "").slice(7).trim();
  const verified = await verifyPopProof(proof, {
    expectedJkt: authed.claims.cnf.jkt,
    htm: req.method,
    htu: `${ctx.config.publicUrl}${path}`,
    accessToken: token,
  });
  if (!ctx.store.consumeJti(verified.jti, "proof", 120)) {
    throw new ToolgateError(ErrorCodes.PROOF_INVALID, "proof replayed");
  }
  return authed;
}

function resolveTool(
  ctx: AppContext,
  tenantId: string,
  upstreamName: string,
  tool: string,
): { upstream: Upstream; toolDef: ToolDef } {
  const upstream = ctx.store.findUpstreamByName(tenantId, upstreamName);
  if (!upstream) throw new ToolgateError(ErrorCodes.NOT_FOUND, `unknown upstream: ${upstreamName}`);
  const toolDef = upstream.tools.find((t) => t.name === tool);
  if (!toolDef) {
    throw new ToolgateError(ErrorCodes.NOT_FOUND, `unknown tool ${tool} on upstream ${upstreamName}`);
  }
  return { upstream, toolDef };
}

function requireApproval(
  ctx: AppContext,
  approvalId: string,
  claims: CapabilityClaims,
): ApprovalRequest {
  const approval = ctx.store.getApproval(approvalId);
  if (!approval) throw new ToolgateError(ErrorCodes.NOT_FOUND, `approval not found: ${approvalId}`);
  if (approval.grantId !== claims.grant_id || approval.tenantId !== claims.tenant) {
    throw new ToolgateError(ErrorCodes.DENIED, "approval belongs to a different grant");
  }
  return approval;
}

interface ExecuteOptions {
  auditBase: {
    id: string;
    tenantId: string;
    ts: string;
    actor: { agentId: string; userId: string; grantId: string; tokenJti: string };
    action: { callId: string; upstream: string; tool: string; argsHash: string };
  };
  decision: AuditDecision;
  grant: DelegationGrant;
  upstream: Upstream;
  toolDef: ToolDef;
  tool: string;
  args: Record<string, unknown>;
}

async function executeCall(ctx: AppContext, o: ExecuteOptions): Promise<unknown> {
  if (!ctx.store.chargeBudget(o.grant.id, o.toolDef.costUnits)) {
    ctx.audit.record({
      ...o.auditBase,
      decision: {
        effect: "deny",
        source: "budget",
        reason: `budget exhausted (cost ${o.toolDef.costUnits})`,
      },
      result: { status: "denied" },
    });
    throw new ToolgateError(ErrorCodes.BUDGET_EXCEEDED, "delegation grant budget exhausted", {
      costUnits: o.toolDef.costUnits,
    });
  }

  const sealed = ctx.store.getSecret(o.upstream.credential.secretRef);
  if (!sealed) throw new ToolgateError(ErrorCodes.INTERNAL, "upstream credential missing from vault");
  const secret = ctx.vault.open(sealed);

  const url = new URL(`${o.upstream.baseUrl.replace(/\/$/, "")}/tools/${o.tool}`);
  const headers: Record<string, string> = { "content-type": "application/json" };
  const cred = o.upstream.credential;
  if (cred.mode === "bearer") headers.authorization = `Bearer ${secret}`;
  else if (cred.mode === "header") headers[cred.headerName.toLowerCase()] = secret;
  else url.searchParams.set(cred.paramName, secret);

  const started = Date.now();
  let response: Response;
  try {
    response = await ctx.fetchImpl(url, {
      method: "POST",
      headers,
      body: JSON.stringify(o.args),
    });
  } catch (err) {
    ctx.audit.record({
      ...o.auditBase,
      decision: o.decision,
      result: { status: "error", latencyMs: Date.now() - started },
    });
    const message = err instanceof Error ? err.message : String(err);
    throw new ToolgateError(ErrorCodes.UPSTREAM_ERROR, `upstream unreachable: ${message}`);
  }

  const latencyMs = Date.now() - started;
  const body: unknown = await response.json().catch(() => ({ raw: "non-JSON upstream response" }));

  if (!response.ok) {
    ctx.audit.record({
      ...o.auditBase,
      decision: o.decision,
      result: { status: "error", httpStatus: response.status, latencyMs },
    });
    throw new ToolgateError(ErrorCodes.UPSTREAM_ERROR, `upstream returned ${response.status}`);
  }

  ctx.audit.record({
    ...o.auditBase,
    decision: o.decision,
    result: {
      status: "executed",
      httpStatus: response.status,
      latencyMs,
      costUnits: o.toolDef.costUnits,
    },
  });
  return body;
}

function toAuditDecision(d: {
  effect: "allow" | "deny" | "require_approval";
  source: "token_bounds" | "rule" | "constraint" | "budget" | "default";
  reason: string;
  ruleId?: string;
}): AuditDecision {
  return {
    effect: d.effect,
    source: d.source,
    reason: d.reason,
    ...(d.ruleId ? { ruleId: d.ruleId } : {}),
  };
}

File: src/toolgate/server/gate.py
import time
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Any

import httpx
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field

from toolgate.core import (
    ApprovalRequest,
    AuditAction,
    AuditActor,
    AuditDecision,
    AuditRecordInput,
    AuditResult,
    CapabilityClaims,
    Decision,
    DelegationGrant,
    ErrorCodes,
    ToolCallContext,
    ToolDef,
    ToolgateError,
    Upstream,
    decide,
    hash_args,
    new_id,
    verify_capability_token,
    verify_pop_proof,
)

from .context import AppContext


class CallBody(BaseModel):
    tool: str = Field(min_length=1)
    args: dict[str, Any] = Field(default_factory=dict)


@dataclass(frozen=True)
class AuthedCall:
    claims: CapabilityClaims
    grant: DelegationGrant


def _now() -> str:
    return datetime.now(UTC).isoformat()


def gate_router(ctx: AppContext) -> APIRouter:
    router = APIRouter(prefix="/v1/gate")

    @router.post("/call/{upstream_name}")
    async def call_tool(upstream_name: str, body: CallBody, request: Request) -> Any:
        authed = await _authenticate(ctx, request, f"/v1/gate/call/{upstream_name}")
        claims, grant = authed.claims, authed.grant
        if not ctx.gate_limiter.allow(grant.id):
            raise ToolgateError(
                ErrorCodes.RATE_LIMITED, "gate call rate limit exceeded for this grant"
            )
        upstream, tool_def = _resolve_tool(ctx, claims.tenant, upstream_name, body.tool)

        call_id = new_id("call")
        call = ToolCallContext(
            upstream=upstream_name, tool=body.tool, args=body.args, cost_units=tool_def.costUnits
        )
        policy = ctx.store.get_policy(grant.policyId)
        if not policy:
            raise ToolgateError(ErrorCodes.INTERNAL, "grant policy missing")

        decision = decide(policy, call, claims.authorization_details)
        actor = AuditActor(
            agentId=claims.act.sub, userId=claims.sub, grantId=grant.id, tokenJti=claims.jti
        )
        action = AuditAction(
            callId=call_id, upstream=upstream_name, tool=body.tool, argsHash=hash_args(body.args)
        )

        if decision.effect == "deny":
            ctx.audit.record(
                AuditRecordInput(
                    id=new_id("evt"),
                    tenantId=claims.tenant,
                    ts=_now(),
                    actor=actor,
                    action=action,
                    decision=_to_audit_decision(decision),
                    result=AuditResult(status="denied"),
                )
            )
            details: dict[str, Any] = {"source": decision.source}
            if decision.rule_id:
                details["ruleId"] = decision.rule_id
            raise ToolgateError(ErrorCodes.DENIED, decision.reason, details)

        if decision.effect == "require_approval":
            approval = ApprovalRequest(
                id=new_id("apr"),
                tenantId=claims.tenant,
                callId=call_id,
                grantId=grant.id,
                agentId=claims.act.sub,
                userId=claims.sub,
                upstream=upstream_name,
                tool=body.tool,
                args=body.args,
                status="pending",
                requestedAt=_now(),
                expiresAt=(
                    datetime.now(UTC) + timedelta(seconds=ctx.config.approval_ttl_seconds)
                ).isoformat(),
            )
            ctx.store.put_approval(approval)
            ctx.store.prune_approvals()
            ctx.audit.record(
                AuditRecordInput(
                    id=new_id("evt"),
                    tenantId=claims.tenant,
                    ts=_now(),
                    actor=actor,
                    action=action,
                    decision=_to_audit_decision(decision),
                    result=AuditResult(status="pending_approval"),
                )
            )
            return JSONResponse(
                status_code=202,
                content={
                    "status": "pending_approval",
                    "approval_id": approval.id,
                    "expires_at": approval.expiresAt,
                    "reason": decision.reason,
                },
            )

        result = await _execute_call(
            ctx,
            actor=actor,
            action=action,
            decision=_to_audit_decision(decision),
            grant=grant,
            upstream=upstream,
            tool_def=tool_def,
            tool=body.tool,
            args=body.args,
        )
        return {"status": "executed", "call_id": call_id, "result": result}

    @router.get("/approvals/{approval_id}")
    async def approval_status(approval_id: str, request: Request) -> dict[str, Any]:
        # Require the PoP proof like every other gate endpoint: a stolen bearer
        # token alone must not be able to enumerate approval state.
        authed = await _authenticate(ctx, request, f"/v1/gate/approvals/{approval_id}")
        approval = _require_approval(ctx, approval_id, authed.claims)
        return {
            "approval_id": approval.id,
            "status": approval.status,
            "expires_at": approval.expiresAt,
        }

    @router.post("/approvals/{approval_id}/execute")
    async def execute_approval(approval_id: str, request: Request) -> dict[str, Any]:
        authed = await _authenticate(ctx, request, f"/v1/gate/approvals/{approval_id}/execute")
        claims, grant = authed.claims, authed.grant
        approval = _require_approval(ctx, approval_id, claims)

        if approval.status == "pending":
            raise ToolgateError(ErrorCodes.APPROVAL_PENDING, "approval still pending")
        if approval.status != "approved":
            raise ToolgateError(ErrorCodes.APPROVAL_DENIED, f"approval is {approval.status}")
        if datetime.fromisoformat(approval.expiresAt) < datetime.now(UTC):
            approval.status = "expired"
            ctx.store.put_approval(approval)
            raise ToolgateError(ErrorCodes.APPROVAL_DENIED, "approval expired before execution")

        # Resolve (read-only) before claiming, so an unknown-upstream error can't
        # strand the claim.
        upstream, tool_def = _resolve_tool(ctx, claims.tenant, approval.upstream, approval.tool)

        # Atomically claim the approval *before* the upstream call. If two
        # requests race, only one wins the approved->executing transition; the
        # loser is rejected instead of firing the side effect a second time.
        if not ctx.store.claim_approval_for_execution(approval.id):
            current = ctx.store.get_approval(approval.id)
            state = current.status if current else "unknown"
            raise ToolgateError(
                ErrorCodes.APPROVAL_DENIED,
                f"approval is {state} (already executing or executed)",
            )

        # The stored args are the approved args — the agent cannot substitute them.
        actor = AuditActor(
            agentId=claims.act.sub, userId=claims.sub, grantId=grant.id, tokenJti=claims.jti
        )
        action = AuditAction(
            callId=approval.callId,
            upstream=approval.upstream,
            tool=approval.tool,
            argsHash=hash_args(approval.args),
        )
        decision = AuditDecision(
            effect="allow",
            source="approval",
            reason=f"approved by {approval.decidedBy or 'unknown'} at {approval.decidedAt or '?'}",
        )

        try:
            result = await _execute_call(
                ctx,
                actor=actor,
                action=action,
                decision=decision,
                grant=grant,
                upstream=upstream,
                tool_def=tool_def,
                tool=approval.tool,
                args=approval.args,
            )
        except Exception:
            # Nothing succeeded upstream — release the claim so a retry is possible.
            ctx.store.revert_approval_claim(approval.id)
            raise

        approval.status = "executed"
        approval.executedAt = _now()
        ctx.store.put_approval(approval)
        return {"status": "executed", "call_id": approval.callId, "result": result}

    return router


# ---------------------------------------------------------------------------
# Pipeline stages
# ---------------------------------------------------------------------------


async def _authenticate_token_only(ctx: AppContext, request: Request) -> AuthedCall:
    auth = request.headers.get("authorization", "")
    if not auth.lower().startswith("bearer "):
        raise ToolgateError(ErrorCodes.TOKEN_INVALID, "missing bearer capability token")
    token = auth[7:].strip()
    claims = verify_capability_token(
        ctx.control_verify_jwk,
        token,
        issuer=ctx.config.issuer,
        audience=ctx.config.gate_audience,
    )

    grant = ctx.store.get_grant(claims.grant_id)
    if not grant:
        raise ToolgateError(ErrorCodes.NOT_FOUND, "grant no longer exists")
    if grant.status != "active":
        # Live tokens die with their grant: revocation is immediate.
        raise ToolgateError(ErrorCodes.REVOKED, "grant revoked")
    if datetime.fromisoformat(grant.expiresAt) < datetime.now(UTC):
        raise ToolgateError(ErrorCodes.TOKEN_EXPIRED, "grant expired")
    agent = ctx.store.get_agent(claims.act.sub)
    if not agent or agent.status != "active":
        raise ToolgateError(ErrorCodes.REVOKED, "agent unknown or disabled")
    return AuthedCall(claims=claims, grant=grant)


async def _authenticate(ctx: AppContext, request: Request, path: str) -> AuthedCall:
    authed = await _authenticate_token_only(ctx, request)
    proof = request.headers.get("x-toolgate-proof")
    if not proof:
        raise ToolgateError(ErrorCodes.PROOF_INVALID, "missing x-toolgate-proof header")

    token = request.headers.get("authorization", "")[7:].strip()
    verified = verify_pop_proof(
        proof,
        expected_jkt=authed.claims.cnf.jkt,
        htm=request.method,
        htu=f"{ctx.config.public_url}{path}",
        access_token=token,
    )
    if not ctx.store.consume_jti(verified.jti, "proof", 120):
        raise ToolgateError(ErrorCodes.PROOF_INVALID, "proof replayed")
    return authed


def _resolve_tool(
    ctx: AppContext, tenant_id: str, upstream_name: str, tool: str
) -> tuple[Upstream, ToolDef]:
    upstream = ctx.store.find_upstream_by_name(tenant_id, upstream_name)
    if not upstream:
        raise ToolgateError(ErrorCodes.NOT_FOUND, f"unknown upstream: {upstream_name}")
    tool_def = next((t for t in upstream.tools if t.name == tool), None)
    if not tool_def:
        raise ToolgateError(
            ErrorCodes.NOT_FOUND, f"unknown tool {tool} on upstream {upstream_name}"
        )
    return upstream, tool_def


def _require_approval(
    ctx: AppContext, approval_id: str, claims: CapabilityClaims
) -> ApprovalRequest:
    approval = ctx.store.get_approval(approval_id)
    if not approval:
        raise ToolgateError(ErrorCodes.NOT_FOUND, f"approval not found: {approval_id}")
    if approval.grantId != claims.grant_id or approval.tenantId != claims.tenant:
        raise ToolgateError(ErrorCodes.DENIED, "approval belongs to a different grant")
    return approval


async def _execute_call(
    ctx: AppContext,
    *,
    actor: AuditActor,
    action: AuditAction,
    decision: AuditDecision,
    grant: DelegationGrant,
    upstream: Upstream,
    tool_def: ToolDef,
    tool: str,
    args: dict[str, Any],
) -> Any:
    def audit(result: AuditResult, dec: AuditDecision = decision) -> None:
        ctx.audit.record(
            AuditRecordInput(
                id=new_id("evt"),
                tenantId=grant.tenantId,
                ts=_now(),
                actor=actor,
                action=action,
                decision=dec,
                result=result,
            )
        )

    if not ctx.store.charge_budget(grant.id, tool_def.costUnits):
        audit(
            AuditResult(status="denied"),
            AuditDecision(
                effect="deny",
                source="budget",
                reason=f"budget exhausted (cost {tool_def.costUnits})",
            ),
        )
        raise ToolgateError(
            ErrorCodes.BUDGET_EXCEEDED,
            "delegation grant budget exhausted",
            {"costUnits": tool_def.costUnits},
        )

    sealed = ctx.store.get_secret(upstream.credential.secretRef)
    if not sealed:
        raise ToolgateError(ErrorCodes.INTERNAL, "upstream credential missing from vault")
    secret = ctx.vault.open(sealed)

    url = f"{upstream.baseUrl.rstrip('/')}/tools/{tool}"
    headers: dict[str, str] = {"content-type": "application/json"}
    params: dict[str, str] = {}
    cred = upstream.credential
    if cred.mode == "bearer":
        headers["authorization"] = f"Bearer {secret}"
    elif cred.mode == "header":
        headers[cred.headerName.lower()] = secret
    else:
        params[cred.paramName] = secret

    started = time.monotonic()
    try:
        response = await ctx.http.post(url, headers=headers, params=params, json=args)
    except httpx.HTTPError as err:
        audit(AuditResult(status="error", latencyMs=(time.monotonic() - started) * 1000))
        raise ToolgateError(ErrorCodes.UPSTREAM_ERROR, f"upstream unreachable: {err}") from err

    latency_ms = (time.monotonic() - started) * 1000
    try:
        body = response.json()
    except ValueError:
        body = {"raw": "non-JSON upstream response"}

    if response.status_code >= 400:
        audit(
            AuditResult(status="error", httpStatus=response.status_code, latencyMs=latency_ms)
        )
        raise ToolgateError(
            ErrorCodes.UPSTREAM_ERROR, f"upstream returned {response.status_code}"
        )

    audit(
        AuditResult(
            status="executed",
            httpStatus=response.status_code,
            latencyMs=latency_ms,
            costUnits=tool_def.costUnits,
        )
    )
    return body


def _to_audit_decision(decision: Decision) -> AuditDecision:
    return AuditDecision(
        effect=decision.effect,
        source=decision.source,
        ruleId=decision.rule_id,
        reason=decision.reason,
    )
